improver/cli/generate_clearsky_solar_radiation.py

```diff
@@ -0,0 +1,90 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to run GenerateClearSkySolarRadiation ancillary generation."""
+from datetime import datetime, timezone
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    target_grid: cli.inputcube,
+    time,
+    accumulation_period: int,
+    *,
+    temporal_spacing=30,
+    altitude: cli.inputcube = 0.0,
+    linke_turbidity: cli.inputcube = 3.0,
+):
+    """Generate a cube containing clearsky solar radiation data, evaluated on the target grid
+    for the specified time and accumulation period. Accumulated clearsky solar radiation is used
+    as an input to the RainForests calibration for rainfall.
+
+    Args:
+        target_grid (iris.cube.Cube):
+            A cube containing the desired spatial grid.
+        time (str):
+            A datetime specified in the format YYYYMMDDTHHMMZ at which to evaluate the
+            accumulated clearsky solar radiation. This time is taken to be the end of
+            the accumulation period.
+        accumulation_period (int):
+            The number of hours over which the solar radiation accumulation is defined.
+        temporal_spacing (int):
+            The temporal spacing between irradiance values used in the evaluation of the
+            accumulated solar radiation, specified in minutes.
+        altitude (iris.cube.Cube):
+            Altitude data used in the evaluation of the clearsky solar irradiance values,
+            specified in metres.
+        linke_turbidity (iris.cube.Cube):
+            Linke turbidity data used in the evaluation of the clearsky solar irradiance
+            values. Linke turbidity is a dimensionless quantity that accounts for the
+            atmospheric scattering of radiation due to aerosols and water vapour, relative
+            to a dry atmosphere.
+
+    Returns:
+        iris.cube.Cube:
+            A cube containing accumulated clearsky solar radiation.
+    """
+    from improver.generate_ancillaries.generate_derived_solar_fields import (
+        GenerateClearskySolarRadiation,
+    )
+
+    time = datetime.strptime(time, "%Y%m%dT%H%MZ").replace(tzinfo=timezone.utc)
```

Comment:
Suggest adding another value_converter for datetimes to avoid duplicating this conversion between the two CLIs in this PR.
It'd also be worth checking if there are any other existing CLIs that are passed datetimes as an argument.

Reply:
I've had a look through all CLIs to see what is done in other cases where datetimes are passed as an argument, and the way they are handled is not entirely consistent:
* [blend_cycles_and_realizations](https://github.com/metoppv/improver/blob/master/improver/cli/blend_cycles_and_realizations.py) - keeps cycletime as `str`. Passes as an argument to process function (and other functions called therein) for `WeightAndBlend` object; eventually converted to datetime using [cycletime_to_datetime](https://github.com/metoppv/improver/blob/e46866908b599d8aecd674e9d03c156560829155/improver/utilities/temporal.py#L58).
* [generate_timezone_mask](https://github.com/metoppv/improver/blob/master/improver/cli/generate_timezone_mask_ancillary.py) - class variable of the same name is initialised with the `str` value (if supplied) when initialising `GenerateTimezoneMask` object; within the process function `_set_time` is called which converts the str to datetime (if not `None`), otherwise tries to extract time from input cube.
* [map_to_timezones](https://github.com/metoppv/improver/blob/master/improver/cli/map_to_timezones.py) - converts time to datetime using [cycletime_to_datetime](https://github.com/metoppv/improver/blob/e46866908b599d8aecd674e9d03c156560829155/improver/utilities/temporal.py#L58) within the CLI process function, then passes the datetime as an argument to `TimezoneExtraction` process function. **(Similar to what is done here.)**
* [temporal_interpolate](https://github.com/metoppv/improver/blob/master/improver/cli/temporal_interpolate.py) - converts a list of `str` values to datetimes. This list of datetimes is then passed as an argument to `TemporalInterpolation` constructor.

I agree that introducing a value_converter would still be worthwhile, however given the variation in how datetime inputs are used, the only other CLI that would utilise this natively would be `map_to_timezones`.

Another question looking across the way datetimes are initialised is whether the associated datetime objects should be timezone aware or naïve? All uses (aside from `generate_timezone_mask` work with timezone naïve datetimes. I think in adding a value_converter, we should probably go with timezone naïve as that seems to be how timezones are handled.

Reply:
> I agree that introducing a value_converter would still be worthwhile, however given the variation in how datetime inputs are used, the only other CLI that would utilise this natively would be `map_to_timezones`.

In fact, even this is not entirely correct: `map_to_timezones` converts local-time to datetime, so the assumed `str` format is `"%Y%m%dT%H%M"`, whereas all other cases take UTC time "%Y%m%dT%H%MZ". A value_converter will need to use a fixed `str` format when doing the datetime conversions, with the latter being the more appropriate choice. So the addition of datetime value converter would currently only be utilised in our usage case.

Reply:
OK, there are a pretty wide variety of ways that string datetimes are currently handled and cleaning all that up is well outside the scope of this PR. I think for the purposes here, leave it as currently is (new value_converter used by these two new CLIs).

It'd be worthwhile creating an issue (on this repo metoppv/improver) regarding this topic so that it can be thought about for future work.

Reply:
I've created issue #1707 regarding this topic.